hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerTask.java

```diff
@@ -162,6 +167,15 @@ public ContainerBalancerTask(StorageContainerManager scm,
    */
   public void run() {
     try {
+      if (delayStart) {
+        long delayDuration = ozoneConfiguration.getTimeDuration(
+            "hdds.scm.wait.time.after.safemode.exit",
+            HddsConfigKeys.HDDS_SCM_WAIT_TIME_AFTER_SAFE_MODE_EXIT_DEFAULT,
+            TimeUnit.SECONDS);
+        LOG.info("ContainerBalancer will sleep for {} seconds before starting" +
+            " balancing.", delayDuration);
+        Thread.sleep(Duration.ofSeconds(delayDuration).toMillis());
```

Comment:
I wonder if we could do something better than sleep here. The sleep means that the tests have to wait for at least the duration of the sleep to pass which makes them slow.

If we have a clock in here, and we change the sleep to `wait()`, then we could advance the clock in a test, and notify the thread to wake it up and check if it should run or not, calculate a new wait time and `wait` again.

Note that #4432 is adding a method to scm to get the clock, so once it is committed we could use it here if we go with this approach.

Reply:
@adoroszlai Do you have any other ideas on how this can be structured to make it more testable and avoid the sleep? Is `wait` the best option?

Reply:
Can you clarify why the tests need to wait? None of the tests need to wait for that duration I think, because we're passing in false for that `delayStart` parameter in tests. All except the `testDelayedStart` test in `TestContainerBalancerTask`, which also doesn't need to wait for sleep to finish because we're just checking if the thread went to sleep and then interrupting it. The rest of the tests are taking a similar duration in this branch compared to master. 

For testing, we could also just check if that log line `ContainerBalancer will sleep for...` is being logged like in `testDelayedStartOnSCMStatusChange`

Reply:
Possible improvement for test:

* extract method `delayStart()`
* override the method in the instance being tested to update an `AtomicBoolean` instead of actual sleep
* assert value of the `AtomicBoolean` depending on whether sleep is expected or not
(or maybe even `AtomicLong` to assert the duration of the sleep if you want to confirm it is using the right config value)

The extracted method also improves `run()`, to keep abstraction at the same level.  (Compare detailed steps of delaying the start to a `balance()` call that hides complexity.)

Reply:
I missed that in the test you were interrupting the thread to wake it up, but then the thread exits, which I guess is OK.

> or maybe even AtomicLong to assert the duration of the sleep if you want to confirm it is using the right config value

That is a good idea - as the current test does not really test that it picks up the correct sleep value, as it just sleeps for a short moment and then gets interrupted.

Reply:
Initially I thought the tests were going to have to wait for the sleep interval before completing, which was wrong. I am happy with the code as it stands now.

Reply:
Unfortunately interruption does not seem to work in all cases.

https://issues.apache.org/jira/browse/HDDS-8676